linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala

```diff
@@ -0,0 +1,611 @@
+          }
+        )
+      else writeType(fieldType.tpe)
+    }
+
+    def writeField(field: StructField): Unit = {
+      b.sameLineList(
+        "field", {
+          appendName(typeID, field.id)
+          writeFieldType(field.fieldType)
+        }
+      )
+    }
+
+    t match {
+      case FunctionType(params, results) =>
+        b.sameLineList(
+          "func", {
+            params.foreach { ty =>
+              b.sameLineList("param", writeType(ty))
+            }
+            results.foreach { ty =>
+              b.sameLineList("result", writeType(ty))
+            }
+          }
+        )
+
+      case ArrayType(fieldType) =>
+        b.sameLineList(
+          "array", {
+            writeFieldType(fieldType)
+          }
+        )
+
+      case StructType(fields) =>
+        b.sameLineList(
+          "struct", {
+            fields.foreach(writeField)
+          }
+        )
+    }
+  }
+
+  private def writeImport(i: Import)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "import", {
+        b.appendElement("\"" + i.module + "\"")
+        b.appendElement("\"" + i.name + "\"")
+
+        i.desc match {
+          case ImportDesc.Func(id, _, typeID) =>
+            b.sameLineList(
+              "func", {
+                appendName(id)
+                writeTypeUse(typeID)
+              }
+            )
+          case ImportDesc.Global(id, _, isMutable, tpe) =>
+            b.sameLineList(
+              "global", {
+                appendName(id)
+                if (isMutable)
+                  b.sameLineList("mut", writeType(tpe))
+                else
+                  writeType(tpe)
+              }
+            )
+          case ImportDesc.Tag(id, _, typeID) =>
+            b.sameLineList(
+              "tag", {
+                appendName(id)
+                writeTypeUse(typeID)
+              }
+            )
+        }
+      }
+    )
+  }
+
+  private def writeSig(params: List[Type], results: List[Type])(
+      implicit b: WatBuilder): Unit = {
+    params.foreach(tpe => b.sameLineList("param", writeType(tpe)))
+    results.foreach(tpe => b.sameLineList("result", writeType(tpe)))
+  }
+
+  private def writeFunction(f: Function)(implicit b: WatBuilder): Unit = {
+    def writeParam(l: Local)(implicit b: WatBuilder): Unit = {
+      b.sameLineList(
+        "param", {
+          appendName(l.id)
+          writeType(l.tpe)
+        }
+      )
+    }
+
+    def writeLocal(l: Local)(implicit b: WatBuilder): Unit = {
+      b.sameLineList(
+        "local", {
+          appendName(l.id)
+          writeType(l.tpe)
+        }
+      )
+    }
+
+    localNames = {
+      val nameGen = new FreshNameGenerator
+      Some((f.params ::: f.locals).map(l => l.id -> nameGen.genName(l.originalName)).toMap)
+    }
+    labelNames = Some(mutable.HashMap.empty)
+    labelNameGen = Some(new FreshNameGenerator)
+
+    b.newLineList(
+      "func", {
+        appendName(f.id)
+        writeTypeUse(f.typeID)
```

Comment:
Nit: IIUC, if we need to distinguish definition and usage, we use `Def` as a suffix for definition and no suffix for usage.

Or maybe here, `writeTypeID` is an appropriate name?